src/DataStructures/DataBox/DataBox.hpp

```diff
@@ -1200,26 +1200,19 @@ static constexpr auto apply(F&& f, const DataBox<BoxTags>& box,
                             tmpl::list<ArgumentTags...> /*meta*/,
                             Args&&... args) noexcept {
   if constexpr (is_apply_callable_v<
-                    F, const const_item_type<ArgumentTags, BoxTags>&...,
-                    Args...>) {
+                    F, decltype(::db::get<ArgumentTags>(box))..., Args...>) {
     return F::apply(::db::get<ArgumentTags>(box)...,
                     std::forward<Args>(args)...);
-  } else if constexpr (::tt::is_callable_v<
-                           std::remove_pointer_t<F>,
-                           tmpl::conditional_t<
-                               std::is_same_v<ArgumentTags, ::Tags::DataBox>,
-                               const DataBox<BoxTags>&,
-                               const_item_type<ArgumentTags, BoxTags>>...,
-                           Args...>) {
+  } else if constexpr (::tt::is_callable_v<std::remove_pointer_t<F>,
+                                           decltype(
+                                               ::db::get<ArgumentTags>(box))...,
+                                           Args...>) {
     return std::forward<F>(f)(::db::get<ArgumentTags>(box)...,
                               std::forward<Args>(args)...);
   } else {
-    error_function_not_callable<
-        std::remove_pointer_t<F>,
-        tmpl::conditional_t<std::is_same_v<ArgumentTags, ::Tags::DataBox>,
-                            const DataBox<BoxTags>&,
-                            const_item_type<ArgumentTags, BoxTags>>...,
```

Comment:
I think the bug is in `const_item_type` because it should be returning a `const T&` but returns a `T`. This is because of the `std::decay_t` on the outside. Why not instead fix that instead? I think this is the reason why we have (but shouldn't need) `const const_item_type<A, B>&` is so many places in stead of `const_item_type<A, B>`.

Note that @markscheel is also fixing another bug in `const_item_type`. The bug there is that because of the use of `decltype` and `declval` a complete type is needed when it shouldn't be. The fix for both basically amounts to:

```cpp
template <typename T>
struct ConvertToConst {
  using type = std::add_const_t<std::add_lvalue_reference_t<T>>;
  // or
  // using type = const T&;
};

template <typename T>
struct ConvertToConst<std::unique_ptr<T>> {
  using type = std::add_const_t<std::add_lvalue_reference_t<T>>;
  // or
  // using type = const T&;
};

template <typename Tag, typename TagList = NoSuchType>
using const_item_type =
    typename ConvertToConst<std::decay_t<storage_type<Tag, TagList>>>;
```
I don't think the interior `std::decay_t` on the last line is needed either, but I'm not sure.

Reply:
Rewrote the implementation to your suggestions 👍

Reply:
Oh, I just saw this.  This partially fixes the same problem as my #2726

Reply:
@markscheel could you remove the `std::decay_t` and take over the `std::optional<std::reference_wrapper>` test in your PR? Then I can close this one.

Reply:
@nilsleiffischer OK I can remove my `std::decay_t`. Do you still want all of your changes in all the other files besides DataBoxTag.hpp?

Reply:
Wait, I get tons of errors when I remove my `std::decay_t`. Maybe I remove my DataBoxTag.hpp changes from my PR and keep the other ones, and you merge your PR?

Reply:
When you remove `std::decay_t` you have to go through the uses of `const_item_type` and remove the `const &` they are wrapped in. I can also do that later, but could be easier if you include that in your PR that touches the files anyway 🤷‍♂️

Reply:
But didn't you already remove all the `const &` stuff in your PR and add the tests?  It's probably easier for me to just remove one commit (my DataBoxTag.hpp) from my PR and merge in mine after yours.

Reply:
It took me maybe 5 minutes :D But sure, I'm happy with just merging this PR quickly and your PR building on it. Could you give it a review then please?